service/src/main/java/bio/terra/axonserver/service/wsm/WorkspaceManagerService.java

```diff
@@ -59,12 +82,85 @@ public ResourceDescription getResource(String accessToken, UUID workspaceId, UUI
    * @throws NotFoundException if workspace does not exist or user does not have access to workspace
    */
   public GcpContext getGcpContext(UUID workspaceId, String accessToken) {
+    return getWorkspace(accessToken, workspaceId, null).getGcpContext();
+  }
+
+  /**
+   * Get the highest IAM role on a workspace for a given user.
+   *
+   * @param accessToken user access token
+   * @param workspaceId terra workspace ID
+   * @return Highest IAM role of the user on the workspace
+   * @throws NotFoundException if workspace does not exist or user does not have access to workspace
+   */
+  public IamRole getHighestRole(String accessToken, UUID workspaceId) {
+    return getWorkspace(accessToken, workspaceId, null).getHighestRole();
+  }
+
+  @VisibleForTesting
+  public AwsCredential getAwsS3StorageFolderCredential(
+      String accessToken,
+      UUID workspaceId,
+      UUID resourceId,
+      AwsCredentialAccessScope accessScope,
+      Integer duration) {
     try {
-      return new WorkspaceApi(getApiClient(accessToken))
-          .getWorkspace(workspaceId, null)
-          .getGcpContext();
-    } catch (ApiException apiException) {
-      throw new NotFoundException("Unable to access workspace " + workspaceId + ".");
+      return new ControlledAwsResourceApi(getApiClient(accessToken))
+          .getAwsS3StorageFolderCredential(workspaceId, resourceId, accessScope, duration);
+    } catch (ApiException e) {
+      throw new NotFoundException("Unable to access workspace or resource.");
+    }
+  }
+
+  /**
+   * Infer the highest level of access that can be requested when obtaining AWS resource
+   * credentials, based on the user's highest IAM role in the workspace
+   *
+   * @param highestRole highest role of user in workspace, can be retrieved via {@link
+   *     #getHighestRole}
+   * @return an access scope that can be used in a call to {@link #getAwsResourceCredential}
+   * @throws ForbiddenException if user does not have at least READER role on workspace
+   */
+  public static AwsCredentialAccessScope inferAwsCredentialAccessScope(
+      @Nullable IamRole highestRole) {
+    if (highestRole == null || highestRole.equals(IamRole.DISCOVERER)) {
```

Comment:
in the current Sam permissions model, APPLICATIONs don't have read or write access to most workspace resources (only the limited set they're managing within the workspace). That might change in the future if needed, but to start I'd rather treat this role like DISCOVERER where they can't read/write the underlying cloud object